File: backend/paths.py
# SPDX-FileCopyrightText: 2026 DennyWanye
# SPDX-License-Identifier: BUSL-1.1

"""Model / asset / user-data directory resolution.

P3-S1 birthed :func:`model_root` and :func:`resolve_model_dir`.
P3-S6 + P3-S7 extend this module to cover the full "where does each kind
of path live" question, because model storage and user data both moved
out of the repo tree when we started shipping a frozen bundle.

P4-S22+ adds **portable mode**: when an ``<install_dir>/userdata/``
folder exists alongside the frozen exe, ALL user-mutable state
(config, SQLite, logs, models, skills, plugins) lives under it
instead of `%AppData%`. This:
  * keeps the C: drive clean (everything next to the install)
  * survives MSI uninstall (Windows Installer only deletes files it
    explicitly listed at install time; an empty userdata/ folder
    sentinel created post-install is preserved)
  * makes "move my install + data to another drive" a directory copy

Detection: the bundle ships an empty ``<install_dir>/userdata/.deskpet-portable``
sentinel. If that sentinel exists, we're portable; if not, classic
AppData layout.

Three orthogonal questions this module answers:

1. **Where are model checkpoints?** (:func:`model_root` / :func:`resolve_model_dir`)
2. **Where does the user's mutable state live?** (:func:`user_data_dir`)
3. **Where do big, regenerable files live?** (:func:`user_cache_dir` /
   :func:`user_models_dir`)

The ``DESKPET_*`` env vars exist so CI, E2E scripts, and dev workflows
can pin paths explicitly without touching the filesystem defaults.
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

import platformdirs

# Constants used by platformdirs. `appauthor=False` means "don't insert
# an AppAuthor\AppName folder stack on Windows" — we own the top-level
# name "deskpet" directly, giving clean paths like `AppData\Roaming\deskpet\`
# instead of `AppData\Roaming\deskpet\deskpet\`.
_APP_NAME = "deskpet"
_APP_AUTHOR: str | bool = False

def _install_dir() -> Path | None:
    """Return the directory containing the frozen exe, or None in dev mode."""
    if getattr(sys, "frozen", False):
        return Path(sys.executable).resolve().parent
    return None


def _portable_userdata_dir() -> Path | None:
    """Return ``<install_root>/userdata/`` when running from a frozen install.

    Layout we expect (Tauri MSI ships this)::

        <install_root>/                  (e.g. G:\\tools\\deskpet\\)
            deskpet.exe                  (Tauri main exe)
            backend/
                deskpet-backend.exe      ← sys.executable when frozen
                _internal/
            userdata/
                config.toml
                data/state.db
                models/
                logs/
                skills/

    The backend exe's ``Path.parent`` is `<install>/backend/`; we walk
    upward to find a sibling ``userdata/`` directory (or its parent
    that *should* contain one). We don't gate on a sentinel file —
    DeskPet has no legacy users to preserve `%AppData%` state for, so
    "frozen ⇒ portable" is the simpler invariant.

    The MSI installer's component manifest does NOT list any files
    under ``userdata/``, so MSI uninstall leaves the entire userdata
    tree (state.db, config.toml, models/, skills/) untouched. That's
    how "uninstall doesn't lose my chat history" works.
    """
    base = _install_dir()
    if base is None:
        return None
    # Pick the "install root" deterministically based on the exe's
    # parent dir name:
    #   exe parent named "backend" → Tauri layout, root is 1 level up
    #   anything else              → standalone, exe parent IS the root
    # Defensive: if the chosen root isn't writable (e.g. Program Files
    # without admin), walk one more level up before giving up.
    if base.name.lower() == "backend":
        candidates_up = (1, 2, 0)
    else:
        candidates_up = (0, 1, 2)
    for steps_up in candidates_up:
        anchor = base
        for _ in range(steps_up):
            anchor = anchor.parent
        candidate = anchor / "userdata"
        try:
            candidate.mkdir(parents=True, exist_ok=True)
        except OSError:
            continue
        # Prove the dir is writable: probe-write then delete.
        probe = candidate / ".deskpet-write-probe"
        try:
            probe.write_bytes(b"")
            probe.unlink()
        except OSError:
            continue
        return candidate
    return None


def is_portable_mode() -> bool:
    """True when running from a frozen install with a writable userdata dir."""
    return _portable_userdata_dir() is not None


def user_data_dir() -> Path:
    """Roaming user data directory (config, DBs, logs).

    Resolution order:
      1. ``DESKPET_USER_DATA_DIR`` env override (tests, power users).
      2. **Portable mode** — ``<install_dir>/userdata/`` (if sentinel
         present). All state lives next to the install, not under C:.
      3. Classic — Windows ``%AppData%\\deskpet\\``, macOS/Linux XDG.
    """
    override = os.environ.get("DESKPET_USER_DATA_DIR")
    if override:
        return Path(override)
    portable = _portable_userdata_dir()
    if portable is not None:
        return portable
    return Path(platformdirs.user_data_dir(_APP_NAME, appauthor=_APP_AUTHOR, roaming=True))


def output_dir(kind: str = "") -> Path:
    """用户生成物统一输出目录: ``<user_data>/OutPut[/<kind>]``。

    桌宠生成的所有交付物(PPT/文档/图片等)都落这里,按类型分子目录
    (如 ``OutPut/PPT``),用户好找。目录按需创建;创建失败返回路径本身
    (调用方写文件时自然报错,不在这里抛)。
    """
    base = user_data_dir() / "OutPut"
    target = base / kind if kind else base
    try:
        target.mkdir(parents=True, exist_ok=True)
    except Exception:  # noqa: BLE001 — 只读盘等;调用方写文件时报错
        pass
    return target


def user_cache_dir() -> Path:
    """Cache directory (HF scratch, temp files).

    Portable: ``<install_dir>/userdata/cache``.
    Classic:  ``%LocalAppData%\\deskpet\\Cache\\`` (Windows).

    Override via ``DESKPET_USER_CACHE_DIR``.
    """
    override = os.environ.get("DESKPET_USER_CACHE_DIR")
    if override:
        return Path(override)
    portable = _portable_userdata_dir()
    if portable is not None:
        return portable / "cache"
    return Path(platformdirs.user_cache_dir(_APP_NAME, appauthor=_APP_AUTHOR))


def user_models_dir() -> Path:
    """Models directory.

    Portable: ``<install_dir>/userdata/models``.
    Classic:  ``%LocalAppData%\\deskpet\\models\\`` (Windows).

    Override via ``DESKPET_MODEL_ROOT`` (same env var :func:`model_root`
    uses, so a single setting covers both lookup paths).
    """
    override = os.environ.get("DESKPET_MODEL_ROOT")
    if override:
        return Path(override)
    portable = _portable_userdata_dir()
    if portable is not None:
        return portable / "models"
    # Sits under LocalAppData\deskpet\ (alongside Cache\, Logs\ if we ever
    # want them) via platformdirs' non-roaming user_data_dir.
    local_base = Path(platformdirs.user_data_dir(_APP_NAME, appauthor=_APP_AUTHOR, roaming=False))
    return local_base / "models"


def user_log_dir() -> Path:
    """User log directory.

    Portable: ``<install_dir>/userdata/logs``.
    Classic:  ``user_data_dir() / "logs"`` (under %AppData% on Windows).

    We deliberately put logs alongside the rest of the user data
    (rather than the platformdirs ``user_log_dir`` default of a
    separate "Logs" subtree) so support bundles capture them in one
    place.

    Override via ``DESKPET_USER_LOG_DIR``.
    """
    override = os.environ.get("DESKPET_USER_LOG_DIR")
    if override:
        return Path(override)
    return user_data_dir() / "logs"


def model_root() -> Path:
    """Return the root directory containing all bundled model subfolders.

    Priority (first hit wins):

    1. ``DESKPET_MODEL_ROOT`` env — explicit override (CI/E2E/debug).
    2. :func:`user_models_dir` if it exists on disk — production path
       post-P3-S6. This is where the installer / setup script drops
       real model weights.
    3. ``sys._MEIPASS/models`` — PyInstaller bundle. Normally empty
       after P3-S6 (too big to fit under P3-G2's 3.5 GB cap), but kept
       as a fallback so tiny bundled checkpoints (or dev-only slim
       bundles) still resolve without extra setup.
    4. Dev fallback: ``backend/models/`` beside this file.

    Note: step 2 requires the directory to **exist** — if the user
    hasn't run ``setup_user_data.ps1`` yet we fall through to the
    bundle / dev path rather than returning a phantom empty dir.
    Callers use :func:`resolve_model_dir` to join a subfolder; they
    decide whether a missing subdir is fatal.
    """
    override = os.environ.get("DESKPET_MODEL_ROOT")
    if override:
        return Path(override)
    # P3-S6: check the production (LocalAppData) path first, but only
    # if it has been provisioned. This lets dev mode keep resolving to
    # backend/models/ without forcing devs to populate LocalAppData.
    user_dir = user_models_dir()
    if user_dir.is_dir():
        return user_dir
    meipass = getattr(sys, "_MEIPASS", None)
    if meipass:
        return Path(meipass) / "models"
    # Dev mode: backend/models/ beside this file.
    return Path(__file__).resolve().parent / "models"


def resolve_model_dir(subdir: str) -> Path:
    """Return absolute path for a named model subfolder, with multi-source fallback.

    ``subdir`` examples: ``"faster-whisper-large-v3-turbo"``, ``"cosyvoice2"``,
    ``"silero_vad"``, ``"bge-m3-int8"``.

    Priority (first **existing** directory wins so user-installed weights
    override the bundle, and bundled weights override "nothing"):

    1. ``DESKPET_MODEL_ROOT/<subdir>`` — explicit env override (CI/E2E).
    2. ``user_models_dir()/<subdir>`` — user-installed (let the user swap
       in their own quantization or fine-tune by dropping it here).
    3. ``sys._MEIPASS/models/<subdir>`` — frozen bundle's bundled weights
       (P4-S20+ install bundle ships BGE-M3 + faster-whisper here).
    4. ``backend/assets/<subdir>`` — dev-mode equivalent of #3 (this is
       what PyInstaller reads from when freezing).
    5. ``backend/models/<subdir>`` — legacy dev-mode location (kept for
       back-compat with pre-P4-S20 dev setups).

    If none of the above exist, returns the path under
    ``user_models_dir()/<subdir>`` so error messages point users at the
    "right" place to drop missing weights. Callers must handle the
    "directory still doesn't exist" case (e.g. faster-whisper falls back
    to HuggingFace auto-download, BGE-M3 falls back to mock vectors).
    """
    override = os.environ.get("DESKPET_MODEL_ROOT")
    if override:
        return (Path(override) / subdir).resolve()

    backend_dir = Path(__file__).resolve().parent
    candidates: list[Path] = [user_models_dir() / subdir]
    meipass = getattr(sys, "_MEIPASS", None)
    if meipass:
        candidates.append(Path(meipass) / "models" / subdir)
    candidates.append(backend_dir / "assets" / subdir)
    candidates.append(backend_dir / "models" / subdir)

    for c in candidates:
        if c.is_dir():
            return c.resolve()

    # Nothing exists → return the user_models_dir path so missing-model
    # errors point at the writable user-friendly location.
    return (user_models_dir() / subdir).resolve()


def ensure_user_dirs() -> None:
    """Create the standard user directories if missing. Idempotent.

    Called once at backend startup (from ``main.py``) so downstream
    code can assume ``user_data_dir() / "data"`` etc. exist without
    each caller racing on ``mkdir(parents=True)``.

    Failures are swallowed with a warning rather than raised — the
    backend should still come up on e.g. a read-only profile, using
    whatever paths the user explicitly configured.
    """
    for d in (
        user_data_dir(),
        user_data_dir() / "data",
        user_data_dir() / "workspace",  # P4-S22: MCP filesystem root
        user_log_dir(),
        user_cache_dir(),
        user_models_dir(),
    ):
        try:
            d.mkdir(parents=True, exist_ok=True)
        except OSError:
            # Non-fatal: user may have overridden to a read-only path.
            # Downstream writes will surface the real error.
            pass

File: tauri-app/src-tauri/src/paths.rs
// SPDX-FileCopyrightText: 2026 DennyWanye
// SPDX-License-Identifier: BUSL-1.1

//! P3-S8 — Rust-side mirror of `backend/paths.py`.
//!
//! Both the Tauri supervisor and the Python backend need to agree on
//! *where* user data lives (`%AppData%\deskpet\`) so the UI can open
//! log / data directories without round-tripping through the backend.
//! We intentionally re-derive the paths here instead of asking the
//! backend: the "open log dir" button must work **even when the backend
//! refuses to start** (that's precisely when the user most needs it).
//!
//! Priority (per env var):
//!   user_data_dir  → `$DESKPET_USER_DATA` || `%AppData%\deskpet`
//!   user_log_dir   → `$DESKPET_USER_LOG`  || `<user_data>\logs`
//!   user_models_dir → `$DESKPET_MODEL_ROOT` || `%LocalAppData%\deskpet\models`
//!
//! The `resolve_*_with` variants accept an env-lookup closure and a
//! pair of base-dir overrides (AppData / LocalAppData) so tests can
//! exercise every branch without mutating real process environment.

use std::path::{Path, PathBuf};

pub type EnvLookup<'a> = &'a dyn Fn(&str) -> Option<String>;

#[derive(Debug, Clone)]
pub struct BaseDirs {
    pub app_data: Option<PathBuf>,      // %AppData% (Roaming)
    pub local_app_data: Option<PathBuf>, // %LocalAppData%
}

impl BaseDirs {
    /// Resolve base dirs from the real OS environment.
    #[cfg(not(test))]
    pub fn from_env() -> Self {
        Self {
            app_data: std::env::var("APPDATA").ok().map(PathBuf::from),
            local_app_data: std::env::var("LOCALAPPDATA").ok().map(PathBuf::from),
        }
    }

    #[cfg(test)]
    pub fn from_env() -> Self {
        // Under cfg(test) we never want to touch real %AppData%.
        Self { app_data: None, local_app_data: None }
    }
}

fn resolve_with(
    env_key: &str,
    env_lookup: EnvLookup<'_>,
    fallback: Option<PathBuf>,
) -> Option<PathBuf> {
    if let Some(v) = env_lookup(env_key).filter(|s| !s.is_empty()) {
        return Some(PathBuf::from(v));
    }
    fallback
}

/// Portable-mode userdata dir: when running from a frozen install, data
/// lives in `<install>/userdata/` next to the exe (mirrors Python
/// `backend/paths.py::_portable_userdata_dir`). Returns None in dev mode
/// (current_exe isn't in the install layout) or when no userdata/ exists.
///
/// Layout: `<install>/deskpet.exe` + `<install>/userdata/`; the backend
/// exe sits at `<install>/backend/deskpet-backend.exe`, so we also check
/// one level up when the exe's parent is named "backend".
#[cfg(not(test))]
fn portable_userdata_dir() -> Option<PathBuf> {
    let exe = std::env::current_exe().ok()?;
    let parent = exe.parent()?;
    let mut anchors: Vec<PathBuf> = vec![parent.to_path_buf()];
    if parent.file_name().and_then(|s| s.to_str()) == Some("backend") {
        if let Some(up) = parent.parent() {
            anchors.insert(0, up.to_path_buf());
        }
    }
    for anchor in anchors {
        let ud = anchor.join("userdata");
        if ud.is_dir() {
            return Some(ud);
        }
    }
    None
}

#[cfg(test)]
fn portable_userdata_dir() -> Option<PathBuf> {
    // 测试下不探测真实 current_exe(避免 test runner 旁碰巧的 userdata 干扰
    // 现有单测的 %AppData% 期望)。
    None
}

pub fn user_data_dir_with(base: &BaseDirs, env_lookup: EnvLookup<'_>) -> Option<PathBuf> {
    // 1. DESKPET_USER_DATA_DIR — 与 Python backend/paths.py 统一的 env 名
    //    (历史 bug: Rust 读 DESKPET_USER_DATA、Python 读 DESKPET_USER_DATA_DIR,
    //    名字不一致 → device_id/onboarding 和 config/db 落到不同目录)。
    if let Some(v) = env_lookup("DESKPET_USER_DATA_DIR").filter(|s| !s.is_empty()) {
        return Some(PathBuf::from(v));
    }
    // 2. 兼容旧 DESKPET_USER_DATA(无 _DIR) — 老用户可能设过,别破坏。
    if let Some(v) = env_lookup("DESKPET_USER_DATA").filter(|s| !s.is_empty()) {
        return Some(PathBuf::from(v));
    }
    // 3. portable: frozen 安装时 <install>/userdata(与 Python portable 一致)。
    if let Some(p) = portable_userdata_dir() {
        return Some(p);
    }
    // 4. classic: %AppData%\deskpet。
    base.app_data.as_ref().map(|p| p.join("deskpet"))
}

pub fn user_log_dir_with(base: &BaseDirs, env_lookup: EnvLookup<'_>) -> Option<PathBuf> {
    if let Some(v) = env_lookup("DESKPET_USER_LOG").filter(|s| !s.is_empty()) {
        return Some(PathBuf::from(v));
    }
    user_data_dir_with(base, env_lookup).map(|p| p.join("logs"))
}

pub fn user_models_dir_with(base: &BaseDirs, env_lookup: EnvLookup<'_>) -> Option<PathBuf> {
    if let Some(v) = env_lookup("DESKPET_MODEL_ROOT").filter(|s| !s.is_empty()) {
        return Some(PathBuf::from(v));
    }
    // portable: <install>/userdata/models(与 Python paths.py 一致)。
    if let Some(p) = portable_userdata_dir() {
        return Some(p.join("models"));
    }
    base.local_app_data.as_ref().map(|p| p.join("deskpet").join("models"))
}

// ---- Public convenience wrappers reading real env ----
// Note: BaseDirs::from_env() returns empty under cfg(test), so these
// helpers safely no-op in unit-test binaries instead of leaking into
// the real %AppData%.

fn real_env(k: &str) -> Option<String> { std::env::var(k).ok() }

pub fn user_data_dir() -> Option<PathBuf> {
    user_data_dir_with(&BaseDirs::from_env(), &real_env)
}

pub fn user_log_dir() -> Option<PathBuf> {
    user_log_dir_with(&BaseDirs::from_env(), &real_env)
}

pub fn user_models_dir() -> Option<PathBuf> {
    user_models_dir_with(&BaseDirs::from_env(), &real_env)
}

/// Ensure `path` exists, creating parent dirs as needed. No-op if Some
/// already points to an existing directory. Returns the path back for
/// chaining in the Tauri command handlers.
#[allow(dead_code)]
pub fn ensure_dir(path: &Path) -> std::io::Result<()> {
    if path.is_dir() {
        return Ok(());
    }
    std::fs::create_dir_all(path)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn env_map(pairs: &'static [(&'static str, &'static str)]) -> impl Fn(&str) -> Option<String> {
        move |k: &str| {
            pairs
                .iter()
                .find(|(kk, _)| *kk == k)
                .map(|(_, v)| v.to_string())
        }
    }

    fn env_empty() -> impl Fn(&str) -> Option<String> {
        |_: &str| None
    }

    fn base_win() -> BaseDirs {
        BaseDirs {
            app_data: Some(PathBuf::from("C:/Users/U/AppData/Roaming")),
            local_app_data: Some(PathBuf::from("C:/Users/U/AppData/Local")),
        }
    }

    #[test]
    fn user_data_dir_defaults_to_appdata_deskpet() {
        let env = env_empty();
        let out = user_data_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("C:/Users/U/AppData/Roaming/deskpet"));
    }

    #[test]
    fn user_data_dir_env_override_wins() {
        let env = env_map(&[("DESKPET_USER_DATA", "D:/custom/deskpet")]);
        let out = user_data_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("D:/custom/deskpet"));
    }

    #[test]
    fn user_data_dir_prefers_dir_suffix_env() {
        // 统一后: DESKPET_USER_DATA_DIR(与 Python 一致)优先于旧 DESKPET_USER_DATA。
        let env = env_map(&[
            ("DESKPET_USER_DATA_DIR", "D:/new/userdata"),
            ("DESKPET_USER_DATA", "E:/old/deskpet"),
        ]);
        let out = user_data_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("D:/new/userdata"));
    }

    #[test]
    fn user_data_dir_falls_back_to_legacy_env() {
        // 只设旧 DESKPET_USER_DATA 时仍兼容(不破坏老用户)。
        let env = env_map(&[("DESKPET_USER_DATA", "E:/old/deskpet")]);
        let out = user_data_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("E:/old/deskpet"));
    }

    #[test]
    fn user_data_dir_empty_env_treated_as_unset() {
        let env = env_map(&[("DESKPET_USER_DATA", "")]);
        let out = user_data_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("C:/Users/U/AppData/Roaming/deskpet"));
    }

    #[test]
    fn user_log_dir_nests_under_user_data() {
        let env = env_empty();
        let out = user_log_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("C:/Users/U/AppData/Roaming/deskpet/logs"));
    }

    #[test]
    fn user_log_dir_independent_env_override() {
        let env = env_map(&[("DESKPET_USER_LOG", "E:/logs")]);
        let out = user_log_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("E:/logs"));
    }

    #[test]
    fn user_models_dir_defaults_to_local_app_data() {
        let env = env_empty();
        let out = user_models_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("C:/Users/U/AppData/Local/deskpet/models"));
    }

    #[test]
    fn user_models_dir_env_override() {
        let env = env_map(&[("DESKPET_MODEL_ROOT", "F:/models")]);
        let out = user_models_dir_with(&base_win(), &env).unwrap();
        assert_eq!(out, PathBuf::from("F:/models"));
    }

    #[test]
    fn missing_app_data_returns_none() {
        let base = BaseDirs { app_data: None, local_app_data: None };
        let env = env_empty();
        assert!(user_data_dir_with(&base, &env).is_none());
        assert!(user_log_dir_with(&base, &env).is_none());
        assert!(user_models_dir_with(&base, &env).is_none());
    }
}
